OIDC: Support OAuth org to role mappings

sathieu opened the pull request:
**What is this feature?**

Ability to map Org to Roles from Oauth token or user info.

Targeting feature-parity between Oauth providers, here is the detail:

  - [x] Social: support AzureAD Oauth org to role mappings ( :x: not needed, AzureAD has no `role_attribute_path` either)
  - [x] Social: support Generic Oauth org to role mappings https://github.com/grafana/grafana/issues/73448
  - [x] Social: support GitHub Oauth org to role mappings https://github.com/grafana/grafana/issues/16763
  - [x] Social: support GitLab Oauth org to role mappings
  - [x] Social: support Google Oauth org to role mappings
  - [x]  Social: support GrafanaCom Oauth org to role mappings ( :x: not needed, GrafanaCom has no `role_attribute_path` either)
  - [x] Social: support Okta Oauth org to role mappings

Quoting from the example in the doc:

> 
> In this example, the user has been granted the role of a `Viewer` in the `org_foo` org, and the role of an `Editor` in the `org_bar` and `org_baz` orgs.
> 
> If the user was a member of the `admin` group, they would be granted the Grafana server administrator role.
> 
> Payload:
> 
> ```json
> {
>     ...
>     "info": {
>         ...
>         "roles": [
>             "org_foo",
>             "org_bar",
>             "another_org'
>         ],
>         ...
>     },
>     ...
> }
> ```
> 
> Config:
> 
> ```ini
> role_attribute_path = contains(info.roles[*], 'admin') && 'GrafanaAdmin' || 'None'
> allow_assign_grafana_admin = true
> org_attribute_path = info.roles
> org_mapping = org_foo:org_foo:Viewer org_bar:org_bar:Editor *:org_baz:Editor
> ```



**Why do we need this feature?**

We are using multiple Grafana installations but want to reduce resource usage. The Org concept allows this but it is currently impossible to automate RBAC from oauth when using multiple Orgs. This PR allows to construct Org to Role mapping using JmesPath.

See also fixed issues (#16763, #73448).

**Who is this feature for?**

Users of Oauth and multiple Orgs.

NB: this feature is already available for LDAP and SAML.

**Which issue(s) does this PR fix?**:

Fixes #16763, #73448

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

sathieu committed:
Social: support GitHub Oauth org to role mappings

Fixes #16763
Signed-off-by: Mathieu Parent <math.parent@gmail.com>

sathieu committed:
Social: support Okta Oauth org to role mappings

Signed-off-by: Mathieu Parent <math.parent@gmail.com>

sathieu committed:
Social: support GitLab Oauth org to role mappings

Signed-off-by: Mathieu Parent <math.parent@gmail.com>

sathieu committed:
Social: support Google Oauth org to role mappings

Signed-off-by: Mathieu Parent <math.parent@gmail.com>

chri2547 commented on docs/sources/setup-grafana/configure-security/configure-authentication/generic-oauth/index.md:
```diff
@@ -197,6 +198,8 @@ If no valid role is found, the user is assigned the role specified by [the `auto
 You can disable this default role assignment by setting `role_attribute_strict = true`.
 This setting denies user access if no role or an invalid role is returned.
 
+The `org_roles_attribute_path` configuration option can be used to add roles to other Orgs. See [Org roles mapping examples]({{< relref "#org-roles-mapping-examples" >}})
```
```suggestion
You can use the `org_roles_attribute_path` configuration option to add roles to other orgs. For more information, refer to [Org roles mapping examples]({{< relref "#org-roles-mapping-examples" >}}).
```

chri2547 commented on docs/sources/setup-grafana/configure-security/configure-authentication/github/index.md:
```diff
@@ -97,6 +98,9 @@ If no valid role is found, the user is assigned the role specified by [the `auto
 You can disable this default role assignment by setting `role_attribute_strict = true`.
 This setting denies user access if no role or an invalid role is returned.
 
+The `org_roles_attribute_path` configuration option can be used to add roles to other Orgs. See [Org roles mapping examples]
```
You can use the `org_roles_attribute_path` configuration option to add roles to other orgs. For more information, refer to [Org roles mapping examples]({{< relref "#org-roles-mapping-examples" >}}).

chri2547 commented on docs/sources/setup-grafana/configure-security/configure-authentication/gitlab/index.md:
```diff
@@ -133,6 +134,8 @@ If no valid role is found, the user is assigned the role specified by [the `auto
 You can disable this default role assignment by setting `role_attribute_strict = true`.
 This setting denies user access if no role or an invalid role is returned.
 
+The `org_roles_attribute_path` configuration option can be used to add roles to other Orgs. See [Org roles mapping examples]({{< relref "#org-roles-mapping-examples" >}})
```
You can use the `org_roles_attribute_path` configuration option to add roles to other orgs. For more information, refer to [Org roles mapping examples]({{< relref "#org-roles-mapping-examples" >}}).

chri2547 reviewed (commented):
Thanks for the contribution! Just a few minor wording suggestions.

sathieu commented:
Thanks @chri2547 for your review :pray:. I've included your suggestions and rebased.

chri2547 reviewed (commented):
Docs are approved. Thank you.

gamab reviewed (commented):
Hey, thanks for the amount of work you put there, great job! FYI, generally speaking, we are not investing in organizations and are working on improving Grafana Teams as an alternative. However, we appreciate the contribution and would be happy to support getting this change merged. Please consider again that we can’t actively support organizations so any feature extension will be on a maintenance mode.

We discussed about this PR with the @grafana/identity-access-team and here are a couple of questions/remarks:
* Using `jmespath` to build the mapping is a cool idea :brain:, but a couple of cases caused me problems, how would you handle them?
  * groups that do not match the organization name:
    ```
    group 'a' maps to org "First.".
    group 'b' maps to org "Second.". 
    ```
  * users are in mutiple groups which have different privileges:
    ```
    users in group 'a' are editors in org 'a'.
    users in group 'b' are viewers in org 'b'. 
    users in both groups are editors in org 'a' AND viewers in org 'b'.
    ```
  * similar to the first and second bullets, working with organization IDs:
    ```
    users in group 'a' are editors in org 2.
    users in group 'b' are viewers in org 3. 
    users in both groups are editors in org 2 AND viewers in org 3.
    ```
* As an alternative, Organization names being changeable and for the sake of consistency with what we did for [saml `org_mapping`](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/saml/#configure-organization-mapping), we could just allow mapping on `OrgID` and have two config options to handle the mapping. 
Example of config:
  ```ini
  ; use groups (in id_token or groups) as an input for the mapping
  org_attribute_path = groups
  ; users in group 'a' will be editors in org 2 and users in group 'b' viewers in org 3. 
  ; all users are viewers in org 1
  org_mapping =  a:2:Editor b:3:Viewer *:1:Viewer
  ```
* We need to enforce the [`role_attribute_strict` config option](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/generic-oauth/#configuration-options). If a user has **no mapping** or has a mapping with an **empty or invalid role**, we should error out.
* If `org_roles_attribute_path` is configured but we found no mapping for the user, they should get the [default role](https://github.com/grafana/grafana/blob/a99485ba331b716af953f02bec4b3325761734f6/pkg/login/social/social.go#L443-L451) in the [default organization](https://github.com/grafana/grafana/blob/a99485ba331b716af953f02bec4b3325761734f6/pkg/services/authn/clients/utils.go#L23-L27). 
* In the edge case where Grafana has been configured with both `role_attribute_path` and `org_roles_attribute_path`, what we did for saml `org_mapping`, is to loop over the `org_mapping` and select the top role between the direct role mapping and the org role mapping. Example:
  ```
  role_attribute_path =mapped_user_to=> Editor
  org_roles_attribute_path =mapped_user_to=> 2: Viewer, 3: Admin
  
  result:
  org_roles = 2: Editor, 3: Admin.
  ```

sathieu commented:
> Hey, thanks for the amount of work you put there, great job! FYI, generally speaking, we are not investing in organizations and are working on improving Grafana Teams as an alternative. However, we appreciate the contribution and would be happy to support getting this change merged. Please consider again that we can’t actively support organizations so any feature extension will be on a maintenance mode.

Thanks @gamab, and thanks for considering this PR.

> We discussed about this PR with the @grafana/identity-access-team and here are a couple of questions/remarks:
> 
>     * Using `jmespath` to build the mapping is a cool idea 🧠, but a couple of cases caused me problems, how would you handle them?

Yes. I think JMespath is very limited, but dynamic mapping is needed. Maybe [go template](https://pkg.go.dev/text/template) generating a YAML would fit. WDYT?

>       
>       * groups that do not match the organization name:
>         ```
>         group 'a' maps to org "First.".
>         group 'b' maps to org "Second.". 
>         ```

Possible with:
```
[{OrgName: 'First.', Role: contains(groups[*], 'a') && 'Editor' || 'None' }, {OrgName: 'Second.', Role: contains(groups[*], 'b') && 'Editor' || 'None' }]
```

>       * users are in mutiple groups which have different privileges:
>         ```
>         users in group 'a' are editors in org 'a'.
>         users in group 'b' are viewers in org 'b'. 
>         users in both groups are editors in org 'a' AND viewers in org 'b'.
>         ```


```
[{OrgName: 'a', Role: contains(groups[*], 'a') && 'Editor' || 'None' }, {OrgName: 'b', Role: contains(groups[*], 'b') && 'Viewer' || 'None' }]
```



>       * similar to the first and second bullets, working with organization IDs:
>         ```
>         users in group 'a' are editors in org 2.
>         users in group 'b' are viewers in org 3. 
>         users in both groups are editors in org 2 AND viewers in org 3.
>         ```

```
[{OrgId: 2, Role: contains(groups[*], 'a') && 'Editor' || 'None' }, {OrgId: 3, Role: contains(groups[*], 'b') && 'Viewer' || 'None' }]
```

>     * As an alternative, Organization names being changeable and for the sake of consistency with what we did for [saml `org_mapping`](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/saml/#configure-organization-mapping), we could just allow mapping on `OrgID` and have two config options to handle the mapping.
>       Example of config:
>       ```ini
>       ; use groups (in id_token or groups) as an input for the mapping
>       org_attribute_path = groups
>       ; users in group 'a' will be editors in org 2 and users in group 'b' viewers in org 3. 
>       ; all users are viewers in org 1
>       org_mapping =  a:2:Editor b:3:Viewer *:1:Viewer
>       ```

That wouldn't fit my use-case, where Orgs come from OIDC claims.

>     * We need to enforce the [`role_attribute_strict` config option](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/generic-oauth/#configuration-options). If a user has **no mapping** or has a mapping with an **empty or invalid role**, we should error out.

Good idea. I can do this by triming any `None` role from OrgRoles, then erroring out when `OrgRoles` is empty and `Role` is empty too.

>     * If `org_roles_attribute_path` is configured but we found no mapping for the user, they should get the [default role](https://github.com/grafana/grafana/blob/a99485ba331b716af953f02bec4b3325761734f6/pkg/login/social/social.go#L443-L451) in the [default organization](https://github.com/grafana/grafana/blob/a99485ba331b716af953f02bec4b3325761734f6/pkg/services/authn/clients/utils.go#L23-L27).

Good idea.

>     * In the edge case where Grafana has been configured with both `role_attribute_path` and `org_roles_attribute_path`, what we did for saml `org_mapping`, is to loop over the `org_mapping` and select the top role between the direct role mapping and the org role mapping. Example:
>       ```
>       role_attribute_path =mapped_user_to=> Editor
>       org_roles_attribute_path =mapped_user_to=> 2: Viewer, 3: Admin
>       
>       result:
>       org_roles = 2: Editor, 3: Admin.
>       ```

I'm OK to use `role_attribute_path` as default role (or `auto_assign_org_role`  if empty), but I think `org_roles_attribute_path` should override this default.

sathieu commented on pkg/services/authn/clients/oauth.go:
```diff
@@ -139,6 +139,11 @@ func (c *OAuth) Authenticate(ctx context.Context, r *authn.Request) (*authn.Iden
 		return userInfo.Role, userInfo.IsGrafanaAdmin, nil
 	})
+	if !c.cfg.OAuthSkipOrgRoleUpdateSync {
+		for orgId, orgRole := range userInfo.OrgRoles {
```
We still use `role_attribute_path` for GrafanaAdmin.

hbjydev commented:
One thing I'd like to point out with this, as Grafana Teams was mentioned as the preferred permissions model now; a lot of companies (like ours) has a team layout that's more akin to business units who should not be able to see each other's data (whether for compliance or security reasons). Because of that, we can't just segregate with teams, because that requires putting datasources for both OUs in the same Grafana tenant, thus having them be able to see each other's data.

That's why this PR is so useful; it allows us to tackle this using multi-tenancy instead of just permissions models that don't work for us.

vtorosyan commented:
@hbjydev thanks for the comment! Indeed the scenario you mentioned is one of the Grafana Teams limitations. We have that in mind and it is one of the first priorities for us to solve that problem. We are currently in a thinking/exploration phase and when things get concrete we will make sure to share with the community to get more feedback and ensure that we won't provide downgraded experience.

gamab commented:
> That wouldn't fit my use-case, where Orgs come from OIDC claims.

What do you mean? What's your `id_token` structure like? 

Maybe we did not understand each others :slightly_smiling_face:. What I'm suggesting is this:
```ini
  ; jmespath to get the list of orgs
  org_attribute_path = path.to.orgs 
  
  ; users in group 'a' will be editors in org 2 and users in group 'b' viewers in org 3. 
  ; all users are viewers in org 1
  org_mapping =  a:2:Editor b:3:Viewer *:1:Viewer
```

So if my `id_token` looks something like this:
```json
{
  "name": "user",
  "organization_memberships": [
    {"name": "a"},
    {"name": "b"}
  ]
}
```

Then I'd configure:
```ini
  ; use groups (in id_token or groups) as an input for the mapping
  org_attribute_path = organization_memberships[*].name
  
  ; users in group 'a' will be editors in org 2 and users in group 'b' viewers in org 3. 
  org_mapping =  a:2:Editor b:3:Viewer
```

> triming any `None` role from `OrgRoles`

Interesting. :thinking: IMO, we'd be better off not using the `None` role for this, in case someone wants explicitly a user to have the `None` role.  
I think you error out if no mapping has been found for the user.

In the previous example, a user that is not a member of organization `"a"` or `"b"` will not be able to login.

sathieu commented:
> > That wouldn't fit my use-case, where Orgs come from OIDC claims.
> 
> What do you mean? What's your `id_token` structure like?

For example with this token:

```json
{
  "viewerGroups": ["org_2", "org_3"],
  "editorGroups": ["org_3", "org_4"]
}
```

The following JmesPath:

```
[viewerGroups[?starts_with(@, 'org_')][{"OrgName": @, "Role": 'Viewer'}],editorGroups[?starts_with(@, 'org_')][{"OrgName": @, "Role": 'Editor'}]][][]
```

Will return:

```json
[
  {
    "OrgName": "org_2",
    "Role": "Viewer"
  },
  {
    "OrgName": "org_3",
    "Role": "Viewer"
  },
  {
    "OrgName": "org_3",
    "Role": "Editor"
  },
  {
    "OrgName": "org_4",
    "Role": "Editor"
  }
]
```

i.e:

- org_2 👁️ Viewer
- org_3 ✍️ Editor
- org_4 ✍️ Editor

> Maybe we did not understand each others 🙂. What I'm suggesting is this:
> 
> ```ini
>   ; jmespath to get the list of orgs
>   org_attribute_path = path.to.orgs 
>   
>   ; users in group 'a' will be editors in org 2 and users in group 'b' viewers in org 3. 
>   ; all users are viewers in org 1
>   org_mapping =  a:2:Editor b:3:Viewer *:1:Viewer
> ```
> 
> So if my `id_token` looks something like this:
> 
> ```json
> {
>   "name": "user",
>   "organization_memberships": [
>     {"name": "a"},
>     {"name": "b"}
>   ]
> }
> ```
> 
> Then I'd configure:
> 
> ```ini
>   ; use groups (in id_token or groups) as an input for the mapping
>   org_attribute_path = organization_memberships[*].name
>   
>   ; users in group 'a' will be editors in org 2 and users in group 'b' viewers in org 3. 
>   org_mapping =  a:2:Editor b:3:Viewer
> ```

Yes. Like [SAML](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/saml/#configure-organization-mapping). The `org_attribute_path` is dynamic, but the `org_mapping` part is static (i.e. in config).


> > triming any `None` role from `OrgRoles`
> 
> Interesting. 🤔 IMO, we'd be better off not using the `None` role for this, in case someone wants explicitly a user to have the `None` role. I think you error out if no mapping has been found for the user.

`None` can be used to workaround a few JMesPath limitations, as shown above (https://github.com/grafana/grafana/pull/77683#issuecomment-1801670859).

> In the previous example, a user that is not a member of organization `"a"` or `"b"` will not be able to login.

Yes. As intended.

sathieu commented on pkg/login/social/social.go:
```diff
@@ -438,6 +454,52 @@ func (s *SocialBase) searchRole(rawJSON []byte, groups []string) (org.RoleType,
+			}
+			orgRoleMapping.OrgID = res.ID
+		} else if orgRoleMapping.OrgID <= 0 {
+			orgRoleMapping.OrgID = 1
```
Done.

sathieu commented on pkg/login/social/social.go:
```diff
@@ -438,6 +454,52 @@ func (s *SocialBase) searchRole(rawJSON []byte, groups []string) (org.RoleType,
+		} else if orgRoleMapping.OrgID <= 0 {
+			orgRoleMapping.OrgID = 1
+		}
+		orgRoles[orgRoleMapping.OrgID] = orgRoleMapping.Role
```
@gamab

> We need to enforce s.roleAttributeStrict there. If true and the role is empty or invalid, we should error out.

Invalid role now errors out (and I've added a test for this).

> If the user is being mapped to different roles in the same org, we should take the highest role.

The latest role wins. I added a test for this.

> If the user has been mapped to a role using the role_attribute_path, then we should take the highest role.

` role_attribute_path`  is for the default org. Using highest role would be harder to understand and to implement.

sathieu commented:
@gamab Please review again :pray:. I've improved the code, and rebased (#77919 created conflicts).

I've also added another example in generic_oauth. Once this looks good for you, I'll add the same kind of doc in all providers.

sathieu commented:
@gamab One more thing: Being `GrafanaAdmin` doesn't give access to non default Orgs. This was already the case before this PR, but I can change the behavior if you want (maybe in another PR, this one is big enough).

gamab commented:
Hey, thanks for working on this again.
I delayed my answer because I wanted to gain some perspective about your decision to create a dynamic `orgRole` mapping using JMESPATH.

Here are our (iam-team) concerns with this dynamic mapping approach:
* it diverges from the saml implementation and we have been working for the past two years on uniformizing providers configurations.
* we also think that it only expresses its true potential, if you can have a structure like yours in the id_token:
  ```json
  {
    "viewerGroups": ["org_2", "org_3"],
    "editorGroups": ["org_3"]
  }
  ```
  
  As then, it's true, you can leverage jmespath to do an `orgRole` mapping.
  ```json
  [
    { "OrgName": "org_2", "Role": "Viewer" },
    { "OrgName": "org_3", "Role": "Viewer" },
    { "OrgName": "org_3", "Role": "Editor" },
  ]
  ```

  But defining such an id_token structure is not possible on all OAuth providers.
  Additionally most of our users would not be able to cover the most common use case:
  ```
  users in group 'a' are editors in org 2.
  users in group 'b' are viewers in org 3. 
  users in both groups are editors in org 2 AND viewers in org 3.
  ```
* it requires expertise from our users to set up the JMESPATH that generates the expected output
* finally, working with org names has three incovenients:
  * It would mean that we need to call the organization service to resolve the user's organizations on every user login
  * Organization names can be changed
  * Organization names must match the IdP ones

As we were saying previously, the only way forward we see for this work would be to follow the `SAML` implementation and have two config options to handle the mapping. 
Example of config:
  ```ini
  ; use groups (in id_token or groups) as an input for the mapping
  org_attribute_path = groups
  ; users in group 'a' will be editors in org 2 and users in group 'b' viewers in org 3. 
  ; all users are viewers in org 1
  org_mapping =  a:2:Editor b:3:Viewer *:1:Viewer
  ```

sathieu commented:
@gamab OK. I'll update this PR to match saml behavior. This will be similar to closed https://github.com/grafana/grafana/pull/23661 by @treydock.

However, the dynamic use-case is still useful. What about configuring this mapping thru the API?

sathieu commented:
@gamab Appart from the dynamic part (which don't affect us yet), the usage of OrgId instead of OrgName is problematic for us. (See related https://github.com/grafana/grafana/issues/77282)

Novakov commented:
I would like to second the opinion of @sathieu - I have an use-case that is nicely solved by using organization names and dynamic mappings of users to organizations.

brandan-schmitz commented:
I also have a use case where it is useful when using an IdP such as Zitadel where you can configure various actions to pass stuff formatted as I want.

Until Grafana Teams can be more useful for separating various things like data for different tenants that should not be able to see each others data through data sources, my only real option is different organizations. But as I often find myself onboarding and offloading various tenants, the process of having to update my Grafana config each time with static org mappings is pretty inconvenient.

gamab commented:
> OrgId instead of OrgName is problematic for us.

@sathieu I looped in our team again. Let's say we are really not big fans of this approach as (for the previously given reasons: adds yet another call to the DB on every user login, adds additional service dependencies, is subject to change, etc...). That being said, you can go on and use the name for the mapping, we will try to make it clear in the docs that this is not the favored way.

> But as I often find myself onboarding and offloading various tenants, the process of having to update my Grafana config each time with static org mappings is pretty inconvenient.

@brandan-schmitz Excellent point, thanks for bringing it up. We have an ongoing project (should be useable by Q4) to make OAuth provider configurations runtime changeable using HTTP endpoints.
I'm assuming that if you create/delete organizations often this is done by a provisioning tool (terraform) or scripting? Once this project is merged, you will be able to modify the Org mapping dynamically as well.

IvanZenger commented:
Hello,

First off, great work. I just wanted to ask what a realistic timeframe is for this PR to be merged and appear in a release?

I work for a tech company and need to migrate from ldap to oauth. The migrate is quite urgent due to security reasons, which is why I am dependent on this PR.

sathieu commented:
> Hello,
> 
> First off, great work. I just wanted to ask what a realistic timeframe is for this PR to be merged and appear in a release?
> 
> I work for a tech company and need to migrate from ldap to oauth. The migrate is quite urgent due to security reasons, which is why I am dependent on this PR.

Hello @IvanZenger. This PR has currently moved down in my TODO-list, because it will take time to rewrite [with the new config settings](https://github.com/grafana/grafana/pull/77683#issuecomment-1824663329), and also because we need to rebuild grafana internally anyway because a few other PRs are blocked or closed:
- #78253
- #76937
- #78560 

Feel free to take-over this task.

IvanZenger commented:
> > OrgId instead of OrgName is problematic for us.
> 
> @sathieu I looped in our team again. Let's say we are really not big fans of this approach as (for the previously given reasons: adds yet another call to the DB on every user login, adds additional service dependencies, is subject to change, etc...). That being said, you can go on and use the name for the mapping, we will try to make it clear in the docs that this is not the favored way.
> 
> > But as I often find myself onboarding and offloading various tenants, the process of having to update my Grafana config each time with static org mappings is pretty inconvenient.
> 
> @brandan-schmitz Excellent point, thanks for bringing it up. We have an ongoing project (should be useable by Q4) to make OAuth provider configurations runtime changeable using HTTP endpoints. I'm assuming that if you create/delete organizations often this is done by a provisioning tool (terraform) or scripting? Once this project is merged, you will be able to modify the Org mapping dynamically as well.

@gamab, if you suggest  "go on and use the name for the mapping"  am I correct in assuming that, for example, the configuration could look like this, using `OrgName` and not being dynamic?


```ini
; use groups (in id_token or groups) as an input for the mapping
org_attribute_path = groups
; users in group 'a' will be editors in org_2 and users in group 'b' viewers in org_3. 
; all users are viewers in org_1
org_mapping =  a:org_2:Editor b:org_3:Viewer *:org_1:Viewer
```

```ini
```

a-patos commented:
Hello,

I am also very interested in this Pull Request. Like Ivan , I understand the importance of migrating from LDAP to OAuth for security reasons. I would like to know if you have obtained any additional information on the merging of this Pull Request and the expected timelines for these changes to be included in an upcoming release?

Looking forward to your response.

gamab commented:
> @gamab, if you suggest "go on and use the name for the mapping" am I correct in assuming that, for example, the configuration could look like this, using OrgName and not being dynamic?

```ini
; use groups (in id_token or groups) as an input for the mapping
 org_attribute_path = groups
; users in group 'a' will be editors in org_2 and users in group 'b' viewers in org_3. 
; all users are viewers in org_1
 org_mapping =  a:org_2:Editor b:org_3:Viewer *:org_1:Viewer
 ```
 
 Yes you are correct :slightly_smiling_face:

IvanZenger commented:
> > @gamab, if you suggest "go on and use the name for the mapping" am I correct in assuming that, for example, the configuration could look like this, using OrgName and not being dynamic?
> 
> 
> 
> ```ini
> 
> ; use groups (in id_token or groups) as an input for the mapping
> 
>  org_attribute_path = groups
> 
> ; users in group 'a' will be editors in org_2 and users in group 'b' viewers in org_3. 
> 
> ; all users are viewers in org_1
> 
>  org_mapping =  a:org_2:Editor b:org_3:Viewer *:org_1:Viewer
> 
>  ```
> 
>  
> 
>  Yes you are correct :slightly_smiling_face: 

Perfect. I already have a running version locally (kinda). Will address this after I’m back from holiday in the second week of January

IvanZenger commented:
I've updated the implementation of the organization role mapping to align with [the new config setting](https://github.com/grafana/grafana/pull/77683#issuecomment-1824663329), resulting in the creation of this PR: https://github.com/sathieu/grafana/pull/3.

I'm seeking feedback and would greatly appreciate a review. Please note, documentation tasks will be addressed following the review process. Thank you in advance for your time. :)

duj4 commented:
hi @IvanZenger ,

We are looking forward to this org mapping function as well to isolate the logs so that user who logged into Grafana will be redirected to its belonged orgs and check logs accordingly.

It seems this PR had been halted for a while and may I know if you have any other workarounds? We are using Generic Oauth.

Thanks.

IvanZenger commented:
> hi @IvanZenger ,
> 
> We are looking forward to this org mapping function as well to isolate the logs so that user who logged into Grafana will be redirected to its belonged orgs and check logs accordingly.
> 
> It seems this PR had been halted for a while and may I know if you have any other workarounds? We are using Generic Oauth.
> 
> Thanks.

Hello @duj4,

Yes. Because of the urgency, I have developed an alternative solution.

I have written a program called `grafana-auth-reverse-proxy`. It authenticates users using OAuth2 and applies the organization mapping during the login process through the Grafana API. Afterwards, it acts as a reverse proxy to Grafana, which is configured with Auth Proxy.

If you're interested, I can look into what it would take to make the project public.

duj4 commented:
> > hi @IvanZenger ,
> > We are looking forward to this org mapping function as well to isolate the logs so that user who logged into Grafana will be redirected to its belonged orgs and check logs accordingly.
> > It seems this PR had been halted for a while and may I know if you have any other workarounds? We are using Generic Oauth.
> > Thanks.
> 
> Hello @duj4,
> 
> Yes. Because of the urgency, I have developed an alternative solution.
> 
> I have written a program called `grafana-auth-reverse-proxy`. It authenticates users using OAuth2 and applies the organization mapping during the login process through the Grafana API. Afterwards, it acts as a reverse proxy to Grafana, which is configured with Auth Proxy.
> 
> If you're interested, I can look into what it would take to make the project public.

Thanks @IvanZenger for the reply, sure, I would like to have a try it in our environment if possible.

titaneric commented:
@IvanZenger I would be very interested to your project! We have the same issue when there are multiple organisations and roles need to apply. We are in the migration from LDAP to oauth2 but we have the current issue.

IvanZenger commented:
> @IvanZenger I would be very interested to your project! We have the same issue when there are multiple organisations and roles need to apply. We are in the migration from LDAP to oauth2 but we have the current issue.

@titaneric @duj4 

I have published the project here: https://github.com/IvanZenger/grafana-auth-reverse-proxy. It works with different IdPs, using OIDC (OAuth2). I had to modify some code to suit the environment and to prevent information leakage. However, I haven't had the opportunity to test these changes or extend the documentation. Contributions are therefore welcome. The Docker release is not yet available.

titaneric commented:
@IvanZenger I could try it locally! Thank you!

a-patos commented:
Hello, so what is missing to validate this Pull Request? I see some administrative work, but I don't see what would be blocking it.

sathieu committed:
Social: support Generic Oauth org to role mappings

Fixes: #73448
Signed-off-by: Mathieu Parent <math.parent@gmail.com>

chri2547 commented:
> @gamab @chri2547 (and others) I've implemented the requested interface, i.e:
> 
> ```ini
> role_attribute_path = contains(info.roles[*], 'admin') && 'GrafanaAdmin' || 'None'
> allow_assign_grafana_admin = true
> org_attribute_path = info.roles
> org_mapping = org_foo:org_foo:Viewer org_bar:org_bar:Editor *:org_baz:Editor
> ```
> 
> Also doc is complete. In short: ready to review !

Tagging @lwandz13 for docs review as this is her domain now. Thank you!

lwandz13 commented on docs/sources/setup-grafana/configure-security/configure-authentication/github/index.md:
```diff
@@ -159,6 +161,25 @@ role_attribute_path = "'Viewer'"
+
+##### Map organization roles
+
+Given a user member of the `@my-github-organization/my-github-team1` and `@my-github-organization/my-github-team2` Gitub teams.
```
```suggestion
Let's use the following Github teams: `@my-github-organization/my-github-team1` and `@my-github-organization/my-github-team2`.
```

lwandz13 replied:
You can also say "Let's say we have the following Github teams: `@my-github-organization/my-github-team1` and `@my-github-organization/my-github-team2`.

lwandz13 commented on docs/sources/setup-grafana/configure-security/configure-authentication/github/index.md:
```diff
@@ -159,6 +161,25 @@ role_attribute_path = "'Viewer'"
+
 ### Example of GitHub configuration in Grafana
 
 This section includes an example of GitHub configuration in the Grafana configuration file.
```
```suggestion
Following is an example of a GitHub configuration in the Grafana configuration file.
```

lwandz13 commented on docs/sources/setup-grafana/configure-security/configure-authentication/gitlab/index.md:
```diff
@@ -251,6 +272,8 @@ The table below describes all GitLab OAuth configuration options. Like any other
 | `auto_login`                 | No       | Set to `true` to enable users to bypass the login screen and automatically log in. This setting is ignored if you configure multiple auth providers to use auto-login.                                                                                                                                                                                              ...
 | `role_attribute_path`        | No       | [JMESPath](http://jmespath.org/examples.html) expression to use for Grafana role lookup. Grafana will first evaluate the expression using the GitLab OAuth token. If no role is found, Grafana creates a JSON data with `groups` key that maps to groups obtained from GitLab's `/oauth/userinfo` endpoint, and evaluates the expression using this data. Finally,  ...
 | `role_attribute_strict`      | No       | Set to `true` to deny user login if the Grafana role cannot be extracted using `role_attribute_path`. For more information on user role mapping, refer to [Configure role mapping]({{< relref "#configure-role-mapping" >}}).                                                                                                                                       ...
+| `org_attribute_path`         | No       | [JMESPath](http://jmespath.org/examples.html) expression to use for Grafana org to role lookup. Grafana will first evaluate the expression using the GitLab OAuth token. If no value is returned, Grafana creates a JSON data with `groups` key that maps to groups obtained from GitLab's `/oauth/userinfo` endpoint, and evaluates the expression using this data ...
```
```suggestion
| `org_attribute_path`         | No       | [JMESPath](http://jmespath.org/examples.html) expression to use for Grafana org to role lookup. Grafana will first evaluate the expression using the GitLab OAuth token. If no value is returned, Grafana creates a JSON data with `groups` key that maps to groups obtained from GitLab's `/oauth/userinfo` endpoint, and evaluates the expression using this data. The result of the evaluation is mapped to org roles based on `org_mapping`. For more information on org to role mapping, refer to [Org roles mapping examples]({{< relref "#org-roles-mapping-examples" >}}).                                                                                                                                                                                                           |                                      |
```

lwandz13 commented on docs/sources/setup-grafana/configure-security/configure-authentication/google/index.md:
```diff
@@ -253,3 +255,22 @@ In this example, all users will be assigned `Viewer` role regardless of the user
+
+##### Map organization roles
+
+Given a user member of the `example-group1@google.com` and `example-group2@google.com` Google groups.
```
This is not a properly structured sentence.  Use something like "Let's say we have the following Google group:  `example-group1@google.com` and `example-group2@google.com`. "  Or "Here are 2 sample Google group": `example-group1@google.com` and `example-group2@google.com`.

lwandz13 commented on docs/sources/setup-grafana/configure-security/configure-authentication/okta/index.md:
```diff
@@ -169,8 +169,29 @@ If no valid role is found, the user is assigned the role specified by [the `auto
+
+##### Map organization roles
+
+Given a user member of the `Group1` and `Group2` groups.
```
Rephrase so this is a proper sentence.

lwandz13 commented on docs/sources/setup-grafana/configure-security/configure-authentication/okta/index.md:
```diff
@@ -169,8 +169,29 @@ If no valid role is found, the user is assigned the role specified by [the `auto
+
 ### Configure team synchronization (Enterprise only)
 
 > **Note:** Available in [Grafana Enterprise]({{< relref "../../../../introduction/grafana-enterprise" >}}) and [Grafana Cloud]({{< relref "../../../../introduction/grafana-cloud" >}}).
```
```suggestion
{{< admonition type="note" >}}
Available in [Grafana Enterprise]({{< relref "../../../../introduction/grafana-enterprise" >}}) and [Grafana Cloud]({{< relref "../../../../introduction/grafana-cloud" >}}).
{{< /admonition >}}
```

sathieu commented:
Thanks @lwandz13 I applied your suggestions. Back to you!

lwandz13 commented:
> Thanks @lwandz13 I applied your suggestions. Back to you!

Thanks for making the updates!

titaneric commented:
Hi, I have known that this PR is relatively large and is ready to be reviewed.

However, I would like to ask does this PR also support multiple orgs and multiple role precedence?
It is hard to describe it in plain English, I add the expected result to belows (borrowed from your testcase).
User's group have `org_foo`, `org_bar`.
And the OrgMapping is `org_foo:11:Viewer, org_bar:11:Editor, *:13:Editor`.
Note that `org_foo` maps to org 11, `org_bar` maps to `11`, but different roles.
Since grafana permission precedence is `GrafanaAdmin` > `Admin` > `Editor` > `Viewer`.
Then the org mapping result becomes `org: 11, role: editor` and `org: 13, role: editor`.
It's probably add new config to support this feature.


```go
		{
			Name:       "org ids",
			OrgMapping: []string{"org_foo:11:Viewer", "org_bar:11:Editor", "*:13:Editor"},
			Groups:     []string{"org_foo", "org_bar", "another_org"},
			ExpectedOrgRoleMapping: []OrgRoleMapping{
				{
					OrgID: 11,
					Role:  "Editor",
				},
				{
					OrgID: 13,
					Role:  "Editor",
				},
			},
		},
```

sathieu commented:
@titaneric This PR implements "last mapping wins",

titaneric commented:
@sathieu ,Thank you! So I have to manipulate the config before running the Grafana if this PR is merged and released.

mgyongyosi commented:
Hi @sathieu, I reviewed and tested this PR thoroughly and there are two issues we need to address before merging this PR:
1. This org mapping logic differs from the SAML org mapping logic. We need to make sure that the org mapping logic is consistent across all the auth providers.
   1. Cannot handle "global org mapping" like `org_foo:*:Editor`.
   1. Handles the wildcard for external groups/roles differently. For example, there is a test called `Given several role claims and valid JMES path, last wins` which I changed and extended the orgMapping with `*:org_foo:Viewer` and the test fails, however the same test passes for SAML:
        ```go
        {
		    name:             "Given several role claims and valid JMES path, last wins",
	 	    orgAttributePath: "info.roles",
		    orgMapping:       []string{"org_foo:org_foo:Viewer", "org_bar:org_foo:Editor", "*:org_foo:Viewer"},
		    rawJSON:          `{"info": {"roles": ["org_foo", "org_bar","another_org"]}}`,
		    expectedOrgRoles: map[int64]roletype.RoleType{11: "Editor"},
      }
        ```
1. We don't want to maintain two different implementations for the same feature. 

However, we appreciate the effort and the contribution you have already made, so we would like to work with you to address these issues to get this PR merged. I extracted the org role mapping logic which is used by SAML to make it reusable for OAuth providers and changed your implementation a bit. [You can find the commit here](https://github.com/grafana/grafana/commit/3123dc4b1d9247c731e664467d0910b881e2da47), please check it out and let me know if you have any questions. I didn't have time to align the tests with the new implementation, so I would appreciate it if you could help with that.

sathieu commented:
@mgyongyosi I'm working on merging your proposal.

> The implementation for AzureAD is missing. AzureAD is a bit different then the rest of the providers, but I think it would make sense to implement it for AzureAD as well.

Can this go to another PR? This one is big already and Azure has no `role_attribute_path` either.

sathieu commented:
@mgyongyosi I've applied the following diff:

```diff
diff --git a/pkg/login/social/connectors/social_base.go b/pkg/login/social/connectors/social_base.go
index 7ebb37b742a..4ee58299237 100644
--- a/pkg/login/social/connectors/social_base.go
+++ b/pkg/login/social/connectors/social_base.go
@@ -178,16 +178,16 @@ func (s *SocialBase) searchRole(rawJSON []byte, groups []string) (org.RoleType,
 
 func (s *SocialBase) extractOrgRoles(ctx context.Context, rawJSON []byte, groups []string, userRole org.RoleType) (map[int64]org.RoleType, error) {
        if s.info.OrgMapping != nil && len(s.info.OrgMapping) > 0 && s.info.OrgAttributePath != "" {
-               groups, err := util.SearchJSONForStringSliceAttr(s.info.OrgAttributePath, rawJSON)
+               orgs, err := util.SearchJSONForStringSliceAttr(s.info.OrgAttributePath, rawJSON)
                if err != nil {
                        return nil, err
                }

-               if len(groups) == 0 {
-                       return nil, nil
+               if len(orgs) == 0 {
+                       orgs = groups
                }

-               return s.orgRoleMapper.MapOrgRoles(groups, s.info.OrgMapping, userRole)
+               return s.orgRoleMapper.MapOrgRoles(orgs, s.info.OrgMapping, userRole)
        }

        return nil, nil
```

Now, tests are passing.

Back to you!

mgyongyosi committed:
wip: Add org_role_mapper tests, refactor connectors

* Connector tests have to be aligned

justinhauer commented:
Any chance this will be going in any time soon?

mgyongyosi committed:
Add more tests after review

mgyongyosi committed:
Working properly with Generic OAuth

mgyongyosi committed:
Add role_attribute_strict handling

mgyongyosi committed:
Update org_role_mapper

mgyongyosi commented:
@sathieu Thanks! We've merged the OrgRoleMapper which is the prerequisites of the org role mapping. 

However, this is a huge PR and reviewing it is quite challenging, in addition this part is sensitive and we'd like to ensure that it's reviewed thoroughly, so we'd like to ask you to split this PR into smaller chunks (one PR per provider) and we can review and merge those one by one. To support this, I've already pushed the Generic OAuth implementation + all the relevant tests that ensures it doesn't change the current behaviour (probably we could add 1-2 other tests and we should also clean up a bit). The Generic OAuth implementation also provides an example on how to implement org role mapping and which tests are mandatory for the rest of the connectors. After the Generic OAuth is merged we can continue with the GitHub provider and so on.

Could you please open a PR for the Generic OAuth connector? Thanks!

sathieu commented:
Thanks @mgyongyosi.

Here it is: https://github.com/grafana/grafana/pull/87394

mgyongyosi commented:
@sathieu Thanks, the Generic OAuth part has been merged. Could you please open the next PR to support org_mapping in the GitHub connector? 🚀

diegommm commented:
Hi! Greetings from the Release Guild. Just letting you know I'm updating your PR's milestone to [11.2.x](https://github.com/grafana/grafana/milestone/529). If you need to see your changes in a previous version, please consider adding a backport label.

Thank you!

mgyongyosi commented:
This feature has been implemented for the OAuth providers in separate PRs.
